src/table.js

```diff
@@ -684,10 +757,112 @@ export function __table(source, operations) {
+
+function coerceRow(object, types, schema) {
+  const coerced = {};
+  for (const col of schema) {
+    const type = types.get(col.name);
+    const value = object[col.name];
+    coerced[col.name] = type === "raw" ? value : coerceToType(value, type);
+  }
+  return coerced;
+}
+
+function initKey() {
+  return {
+    other: 0,
+    boolean: 0,
+    integer: 0,
+    number: 0,
+    date: 0,
+    string: 0,
+    array: 0,
+    object: 0,
+    bigint: 0,
+    buffer: 0
+  };
+}
+
+// We need to show *all* keys present in the array of Objects
+function getAllKeys(rows) {
+  const keys = new Set();
+  for (const row of rows) {
+    // avoid crash if row is null or undefined
+    if (row) {
+      // only enumerable properties
+      for (const key in row) {
+        // only own properties
+        if (Object.prototype.hasOwnProperty.call(row, key)) {
+          // unique properties, in the order they appear
+          keys.add(key);
+        }
+      }
+    }
+  }
+  return Array.from(keys);
+}
+
+export function inferSchema(source, columns = getAllKeys(source)) {
+  const schema = [];
+  const sampleSize = 100;
+  let sample = source.slice(0, sampleSize);
+  const typeCounts = {};
+  for (const d of sample) {
+    for (const col of columns) {
+      if (!typeCounts[col]) typeCounts[col] = initKey();
+      const type = typeof d[col];
+      const value = type === "string" ? d[col].trim() : d[col];
+      typeCounts[col][inferType(type, value)]++;
+    }
+  }
+  for (const col in typeCounts) {
+    let type = greatest(Object.keys(typeCounts[col]), d => typeCounts[col][d]);
+    const numDefined = sample.filter(d => !(d[col] == null || d[col] === "")).length;
+    // If over 90% of the sampled data counted as this type, use it. Otherwise,
+    // use "other."
+    type = typeCounts[col][type] / numDefined >= 0.9 ? type : "other";
+    schema.push({
+      name: col,
+      type: type,
+      inferred: type
+    });
+  }
+  return schema;
+}
+
+function inferType(type, value) {
+  // for json and sqlite, we already have some types, but for csv and tsv, all
+  // columns are strings here.
+  const typedNonStrings = ["bigint", "boolean", "object"];
+  if (type !== "string") {
+    if (Array.isArray(value)) return "array";
+    else if (value instanceof Date) return "date";
+    else if (value instanceof ArrayBuffer) return "buffer";
+    else if (type === "number") {
+      if (Number.isInteger(+value)) return "integer";
+      else return "number";
+    }
+    else if (typedNonStrings.includes(type)) return type;
+    else return "other";
+  } else {
+    if (value === "true" || value === "false") return "boolean";
+    else if (value && !isNaN(value)) {
+      if (Number.isInteger(+value)) return "integer";
+      else return "number";
+    } else if (/^\d+n$/.test(value)) return "bigint";
+    else if (
+      /^(([-+]\d{2})?\d{4}(-\d{2}(-\d{2}))|(\d{1,2})\/(\d{1,2})\/(\d{2,4}))([T ]\d{2}:\d{2}(:\d{2}(\.\d{3})?)?(Z|[-+]\d{2}:\d{2})?)?$/.test(value)
```

Comment:
Can we pull this regex out to a constant so that we don’t need to repeat it twice?